### PR TITLE
[第三組] Refactor/error handling

### PR DESCRIPTION
**Pull request by percykugood**
### Purpose
This pull request introduces significant updates to the codebase, including enhancements to the CI workflow, the addition of a new crawler implementation, and improvements to the exception handling and LLM client modules.

### Exception Handling:
* Introduced a base class `APIException` in `backend/src/exception_handler/base.py` for standardized API exception handling.
* Added specific exception classes in `backend/src/exception_handler/exception.py` for various error scenarios, such as `NoResourceFoundException`, `AuthenticationExceptionBase`, and its subclasses.

### Configuration Updates:
* Updated `SENTRY_DSN` in `backend/src/config.py` to use a new DSN for error tracking.

### CODE REVIEW

**Comment by seanpai96** [comment]
Please fix your conflict so I can merge!
Also your CI didn't pass.

**Comment by yozen0405** [comment]
這次更新顯示了卓越的技術實力，從 CI 工作流程的強化到新增爬蟲實作及改進例外處理與 LLM 客戶端模組，都展現了對系統穩定性與可擴展性的高度重視。標準化的 API 例外處理和精細的錯誤追蹤設定，為系統帶來了更高的健壯性與可靠性，值得讚賞。

**Comment by EdwardKYC** [comment]
從 CI 工作流程的強化到新增爬蟲實作及改進例外處理與 LLM 客戶端模組，都展現了對系統穩定性與可擴展性的高度重視。為系統帶來了更高的健壯性與可靠性。但...code跟coffee高度重疊欸，而且有conflict。然後github上的ci test也沒過啊，新增了很多函式但感覺沒用到太多 重複性很高。

**Comment by Rogue8464** [comment]
此次更新強調用logger和錯誤處理來記錄問題以增加系統穩定性，但存在CI 測試未通過及高函式重複性等問題，需進一步優化與整合。

**Comment by ptchen314** [comment]
老兄的Exception做的很好，可以直觀看到錯誤內容，但你CI沒過ㄟQQ

**Comment by mlcp29th-hi** [comment]
Thanks for your work on this PR! I noticed some areas, particularly in the integration parts of the crawler, where error handling could be enhanced. Here are my suggestions:

1. **Handling NoneType Errors**: 
   - The error logs indicate that there was an issue with handling `NoneType` objects in your login functionality. It's essential to have checks for `None` before accessing attributes. This can prevent runtime exceptions and provide more informative error messages.
   - Consider adding a guard clause in your login code to check if the user object is `None`:
     ```python
     if user is None:
         raise ValueError("User not found or not logged in.")
     ```

2. **DomainMismatchException**: 
   - In the `validate_and_parse` method, when raising a `DomainMismatchException`, it would be helpful to log the offending URL. This would make diagnosing issues easier in the future.
   - Example:
     ```python
     if not self._is_valid_url(url):
         logger.error(f"Domain mismatch for URL: {url}")
         raise DomainMismatchException(url)
     ```

3. **Error Handling in HTTP Requests**:
   - Ensure that HTTP requests in your Crawler classes handle potential exceptions such as `requests.exceptions.RequestException`. Wrapping the request call in a try-catch block can improve the robustness of your application.
   - For example:
     ```python
     try:
         response = requests.get(url, timeout=self.timeout)
         response.raise_for_status()  # Raise an error for HTTP error responses
     except requests.exceptions.RequestException as e:
         logger.error(f"HTTP request failed: {e}")
         return []
     ```

4. **Save Method Error Handling**: 
   - In the `save` method for saving news articles, it's advisable to handle potential database errors. You could use a try-except block to catch exceptions related to database transactions.
   - Example:
     ```python
     def save(self, news: News, db: Session | None):
         try:
             db.add(news)
             db.commit()  # Make sure to catch any commit issues
         except Exception as e:
             logger.error(f"Failed to save news article: {e}")
             db.rollback()  # Rollback in case of an error
     ```

5. **Logging**: 
   - Utilizing a combination of error logging and user-friendly messages helps both development and support teams understand what went wrong without diving too deep into stack traces.
   - Ensure that all significant points of failure in your workflow (like fetching headlines, parsing, and database operations) are logged.

By enhancing error handling in these ways, we can improve not only the stability of the application but also make troubleshooting easier in the long run. Looking forward to seeing these adjustments!

**Comment by jxuanyang** [comment]
CI沒過但其他頗好

**Comment by Yuannzh1** [comment]
### 優點

1. **新增的 Try-Except 功能**：
   - **異常捕捉範圍明確**：針對特定的異常類型進行處理，如 `SQLAlchemyError`、`ExpiredSignatureError` 等，有助於提高異常處理的精確性。
   - **詳細的日誌記錄**：使用 `logger.error` 記錄詳細的異常訊息，有助於後續的錯誤追蹤和分析。
   - **自訂異常類別**：定義了多個自訂異常類別，如 `NoResourceFoundException`、`ArticleNotFoundException` 等，這些異常類別名稱直觀，能夠清楚地表達異常的具體情況。

### 改進建議

1. **Try-Except 內是否處理太多事情**：
   - **細化 Try-Except 塊**：在某些情況下，Try-Except 塊內處理的事情過多，建議將其細化。例如在 `router.py` 中的 `search_news` 函數內，可以將不同的步驟分別放在各自的 Try-Except 塊中，以提高可讀性和可維護性。
   ```python
   try:
       response = requests.get(news["titleLink"])
       soup = BeautifulSoup(response.text, "html.parser")
   except Exception as e:
       logger.error(f"Error fetching news content: {str(e)}")
       raise HTTPException(status_code=500, detail="Error fetching news content")

   try:
       title = soup.find("h1", class_="article-content__title").text
       time = soup.find("time", class_="article-content__time").text
       content_section = soup.find("section", class_="article-content__editor")
       paragraphs = [p.text for p in content_section.find_all("p") if p.text.strip() != "" and "▪" not in p.text]
   except Exception as e:
       logger.error(f"Error parsing news content: {str(e)}")
       raise HTTPException(status_code=500, detail="Error parsing news content")
   ```

總結來說，這次改動在異常處理、日誌記錄和自訂異常的設計上做得很好，但可以通過進一步細化 Try-Except 塊、增加更多的日誌記錄和統一異常處理方式來進一步提升程式碼的可讀性和可維護性。

**Comment by Diego0923** [comment]
改動增加了系統穩定性，可惜CI 測試未通過，
建議可以優化錯誤處理及日誌分級，避免高重複性函式；
細化異常類型並統一處理方式，提升穩定性與可維護性。

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -6,10 +6,14 @@ on:
...
       - 'develop'
       - 'fix/integration-test'
+      - 'refactor/crawler-oop-implementation'
+      - 'refactor/llm_client'
+      - 'refactor/error_handling'
   pull_request:
     branches:
       - 'main'
       - 'develop'
+      - 'refactor/error_handling'
```
Comment (dee9c-git) [inline .github/workflows/ci.yaml:16]: I don't think you will pull into error_handling

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -39,7 +44,8 @@ jobs:
 
       - name: Run integration tests (with mock DB)
         env:
-          OPENAI_API_KEY: ${{ secrets.OPENAI_API_KEY }}
+          openai: ${{ secrets.OPENAI }}
+          claude: ${{ secrets.CLAUDE }}
```
Comment (dee9c-git) [inline .github/workflows/ci.yaml]: I recommend still using ```OPENAI_API_KEY: ${{ secrets.OPENAI_API_KEY }}``` or ```openai_token``` at least.
openai doesn't tell you what it is, it's an api key.

File: backend/app.log
Comment (dee9c-git) [inline backend/app.log:1]: git ignore it

File: backend/app_rotating.log
Comment (dee9c-git) [inline backend/app_rotating.log:1]: git ignore it

File: backend/src/exception_handler/base.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        """
+        Get the error message for this exception.
+
+        Returns:
+            str: The error message to display in the response or logs.
+
+        Note:
+            This method must be implemented by subclasses.
+        """
+        pass
```
Comment (dee9c-git) [inline backend/src/exception_handler/base.py:44]: Is it possible to use ```return NotImplemented ``` to make the note enforced?

File: backend/src/llm_client/base.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
+from abc import ABC, abstractmethod
+from pydantic import BaseModel, Field
+import aisuite as ai
+client = ai.Client()
+
+models = ["openai:gpt-3.5-turbo", "anthropic:claude-3-5-sonnet-20240620"]
```
Comment (dee9c-git) [inline backend/src/llm_client/base.py]: I don't think this is good.
A simply change in order, ```models = ["anthropic:claude-3-5-sonnet-20240620", "openai:gpt-3.5-turbo"]```
and hell breaks loose

File: backend/src/news/utils.py [hunk 5]
```diff
@@ -39,33 +36,10 @@ def get_new_info(search_term, fetch_all_pages=False):
...
-            "id": f"search:{quote(search_term)}",
-            "channelId": 2,
-            "type": "searchword",
-        }
-        response = requests.get("https://udn.com/api/more", params=params)
+    
 
-        news_data = response.json()["lists"]
-    return news_data
+    headlines = crawler.get_headline(search_term, page=(1, 10) if fetch_all_pages else 1)
```
Comment (dee9c-git) [inline backend/src/news/utils.py]: Magic numbers perhaps?

File: backend/src/logger/logger.py
Comment (dee9c-git) [inline backend/src/logger/logger.py:1]: Why exactly is this in a folder?

**Review by dee9c-git — COMMENTED** [review]
Little problems here and there, but I think they are all small and manageable.
Bit of fixing and you're good to go!

File: backend/app.log
Comment (kevin0216) [inline backend/app.log:1]: It's strongly suggested that log file should not uploaded to GitHub repository, it should be ignored (Add *.log into .gitignore to ignore it)

File: backend/src/news/router.py [hunk 6]
```diff
@@ -106,42 +138,38 @@ def upvote_article(
...
+            response["summary"] = result["影響"]
+            response["reason"] = result["原因"]
+        return response
+    except Exception as e:
+        logger.error(f"Error processing article: {str(e)}")
+        raise HTTPException(status_code=500, detail="Error processing article")
+
+@router.get("/sentry-debug")
+async def trigger_error():
+    division_by_zero = 1 / 0
```
Comment (kevin0216) [inline backend/src/news/router.py:174]: This part should be removed after local testing

File: backend/src/user/router.py [hunk 7]
```diff
@@ -32,11 +33,17 @@ async def login_for_access_token(
...
     """
-    user = check_user_password_is_correct(db, form_data.username, form_data.password)
-    access_token = create_access_token(
-        data={"sub": str(user.username)}, expires_delta=timedelta(minutes=AppConfig.INTERVAL_MINUTES)
-    )
-    return {"access_token": access_token, "token_type": "bearer"}
+    try:
+        user = check_user_password_is_correct(db, form_data.username, form_data.password)
+        if not user:
+            logger.warning(f"Failed to login: Invalid credentials")
```
Comment (kevin0216) [inline backend/src/user/router.py:39]: The user is missing here, then you should just return a `HTTPException` and stop the function, as access token won't be successfully created with missing User object (expected Exception)
```py
 if not user:
      logger.warning(f"Failed to login: Invalid credentials")
      raise HTTPException(status_code=401, detail="User not found") # or other Exception class you defined
```

File: backend/src/user/router.py [hunk 8]
```diff
@@ -51,11 +58,16 @@ def create_user(user: UserAuthSchema, db: Session = Depends(session_opener)):
...
-    db.add(db_user)
-    db.commit()
-    db.refresh(db_user)
-    return db_user
+    try:
+        db_user = User(username=user.username, hashed_password=hashed_password)
+        db.add(db_user)
+        db.commit()
+        db.refresh(db_user)
+        return db_user
```
Comment (kevin0216) [inline backend/src/user/router.py:66]: `db.commit()` is the only function that will throw exception, you can just catch this function, rest of them are not required to do so.

File: backend/src/exception_handler/exception.py
Comment (kevin0216) [inline backend/src/exception_handler/exception.py:1]: I didn't see anywhere these Exception classes are being used?
What is the purpose of these Exception classes being defined?

**Pull request merged**